interfaces/builtin/wayland.go

```diff
@@ -19,32 +19,127 @@
 
 package builtin
 
+import (
+	"github.com/snapcore/snapd/interfaces"
+	"github.com/snapcore/snapd/interfaces/apparmor"
+	"github.com/snapcore/snapd/interfaces/seccomp"
+	"github.com/snapcore/snapd/interfaces/udev"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
 const waylandSummary = `allows access to compositors supporting wayland protocol`
 
 const waylandBaseDeclarationSlots = `
   wayland:
     allow-installation:
       slot-snap-type:
+        - app
         - core
+    deny-auto-connection:
+      on-classic: false
 `
 
-const waylandConnectedPlugAppArmor = `
-# Description: Can access compositors supporting the wayland protocol
+const waylandPermanentSlotAppArmor = `
+# Description: Allow operating as a Wayland display server. This gives privileged access
+# to the system.
+
+# needed since Wayland is a display server and needs to configure tty devices
+capability sys_tty_config,
+/dev/tty[0-9]* rw,
+
+# Create the Wayland socket and lock file
+/run/user/[0-9]*/wayland-[0-9]* rw,
```

Comment:
I think you will need:
```
/run/user/[0-9]*/wayland-shared-* rw,
/run/user/[0-9]*/wayland-cursor-shared-* rw,
```

Reply:
I've been tracking these down, these are files for shared memory created by wayland  (another one being xwayland-shared), to pass data between client and compositor, for example pass pixel buffers, cursor images.

The tricky thing is that clients will create them in their XDG_RUNTIME_DIR. For a snapped app, this directory is per-snap. Non-snapped apps use the normal /run/user/`id -u`

It is correct that the Wayland socket is in a shared place, so multiple clients can connect to it. But these shared memory files I think can be confined inside the snap's XDG_RUNTIME_DIR, as I don't believe they are ever shared between clients, they are purely for client-server communication.

For snapped apps, I think in addition to your proposal we should add

      /run/user/[0-9]*/snap.*/wayland-shared-* rw,
      /run/user/[0-9]*/snap.*/wayland-cursor-shared-* rw,
      /run/user/[0-9]*/snap.*/xwayland-shared-* rw,

(XWayland clients create the later). WDYT?

Reply:
It would be fine for the server to do this, but note that @jhenstridge is working on user mounts. This will allow the snap's XDG_RUNTIME_DIR to be /run/user/`id -u` on the inside, but bind mounted on top of a snap-specific directory, then snapd/snap-confine/snap-update-ns would bind mount the wayland socket into the snap-specific dir. This would mean we could get rid of the desktop helper bits for the compat symlink (this also more closely follows the upstream designs). 

How these intersect with the above accesses... I'm not sure. With blackbox testing, it seems to be initiated on the client side (they do an open() then unlink()) in the client's XDG_RUNTIME_DIR and I think the client must pass the fd over to the server. Ie, if I add an appropriate 'audit deny' rule, I can see denials like this:

```
apparmor="DENIED" operation="mknod" profile="snap.gedit.gedit" name="/run/user/1000/snap.gedit/wayland-cursor-shared-Hcwp8Q" pid=4896 comm="gedit" requested_mask="c" denied_mask="c" fsuid=1000 ouid=1000
```

as such, what you should do is add waylandConnectedSlotAppArmor policy, like so:
```
const waylandConnectedSlotAppArmor = `
# Allow access to client Wayland sockets for connected snaps
owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/wayland-shared-* rw,
owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/wayland-cursor-shared-* rw,
owner /run/user/[0-9]*/###PLUG_SECURITY_TAGS###/xwayland-shared-* rw,
```

This will allow the server to access these sockets for connected snaps. If my blackbox testing analysis is correct, this should continue to work after user mounts are in effect, because the slot side will continue to see these in /run/user/`id -u`/snap.*.

Reply:
In addition to the above, the waylandPermanentSlotAppArmor policy would need this to handle unconfined applications:
```
owner /run/user/[0-9]*/wayland-shared-* rw,
owner /run/user/[0-9]*/wayland-cursor-shared-* rw,
owner /run/user/[0-9]*/xwayland-shared-* rw,
```